internal/handlers/jsonb1/msg_aggregate.go

```diff
@@ -0,0 +1,194 @@
+	"github.com/FerretDB/FerretDB/internal/wire"
+)
+
+// MsgAggregate executes an aggregation pipeline on documents in a collection.
+func (h *storage) MsgAggregate(ctx context.Context, msg *wire.OpMsg) (*wire.OpMsg, error) {
+	document, err := msg.Document()
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	m := document.Map()
+	pipeline := m["pipeline"].(types.Array)
+	db := m["$db"].(string)
+	collection := m["aggregate"].(string)
+
+	sql, args, err := aggregateToSQL(pipeline, db, collection)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	rows, err := h.pgPool.Query(ctx, sql, args...)
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+	defer rows.Close()
+
+	var docs types.Array
+
+	for {
+		doc, err := nextRow(rows)
+		if err != nil {
+			return nil, lazyerrors.Error(err)
+		}
+		if doc == nil {
+			break
+		}
+
+		docs = append(docs, *doc)
+	}
+
+	var reply wire.OpMsg
+	err = reply.SetSections(wire.OpMsgSection{
+		Documents: []types.Document{types.MustMakeDocument(
+			"cursor", types.MustMakeDocument(
+				"firstBatch", docs,
+				"id", int64(0), // TODO
+				"ns", db+"."+collection,
+			),
+			"ok", float64(1),
+		)},
+	})
+	if err != nil {
+		return nil, lazyerrors.Error(err)
+	}
+
+	return &reply, nil
+}
+
+func aggregateToSQL(pipeline types.Array, db, collection string) (string, []interface{}, error) {
+	filterPairs := []interface{}{}
+	groupByFields := map[string]string{}
+	sortByFields := map[string]interface{}{}
+	aggregations := map[string]string{}
+	for _, item := range pipeline {
+		if p, ok := item.(types.Document); ok {
+			if p.Map()["$match"] != nil {
+				tfilter := p.Map()["$match"].(types.Document)
+				for k, v := range tfilter.Map() {
+					filterPairs = append(filterPairs, k)
+					filterPairs = append(filterPairs, v)
+				}
+			}
+			if p.Map()["$sort"] != nil {
+				tsort := p.Map()["$sort"].(types.Document)
+				for k, v := range tsort.Map() {
+					sortByFields[k] = v
+				}
+			}
+			if p.Map()["$group"] != nil {
+				tgroup := p.Map()["$group"].(types.Document)
+				for k, v := range tgroup.Map() {
+					if fieldAgg, ok := v.(types.Document); ok {
+						for agg, val := range fieldAgg.Map() {
+							aggFormat := "count(%s)"
+							aggValue := "*"
+							if agg == "$sum" {
+								aggFormat = "coalesce(sum(%s),0)"
+								aggValue = "1"
+							}
+							if agg == "$sortByCount" {
+								aggFormat = "count(%s)"
+							}
+							if sval, ok := val.(string); ok {
+								aggValue = fmt.Sprintf("_jsonb->'%s')", sval)
+							}
+							aggregations[k] = fmt.Sprintf(aggFormat, aggValue)
+						}
+					}
+					if fieldName, ok := v.(string); ok {
+						groupByFields[k] = fmt.Sprintf("_jsonb->'%s'", strings.TrimPrefix(fieldName, "$"))
+					}
+				}
+			}
+		}
+	}
+
+	filter, err := types.MakeDocument(filterPairs...)
+	if err != nil {
+		return "", nil, lazyerrors.Error(err)
+	}
+
+	groupByFieldItems := []string{}
+	groupByClauseItems := []string{}
```

Comment:
```suggestion
	groupByFieldItems := make([]string{}, 0, len(groupByFields))
	groupByClauseItems := make([]string{}, 0, len(groupByFields))
```

Let's allocate the whole slice where we know the capacity. In other places too